fix(semantic): resolve reference incorrectly when a parameter references a parameter that hasn't been defined yet

Dunqing opened the pull request:
close: #3682

The TypeScript code that handles this is [here](https://github.com/Dunqing/TypeScript/blob/d8086f14b6b97c0df34a0cc2f56d4b5926a0c299/src/compiler/utilities.ts#L11515-L11577). It looks complicated.

Dunqing commented:
* **#4004** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4004?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Dunqing and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

Boshen reviewed (approved):
This is indeed more logical than what we had before.

Dunqing committed:
fix(semantic): resolve reference incorrectly when a parameter references a parameter that hasn't been defined yet (#4004)

close: #3682

The TypeScript code that handles this is [here](https://github.com/Dunqing/TypeScript/blob/d8086f14b6b97c0df34a0cc2f56d4b5926a0c299/src/compiler/utilities.ts#L11515-L11577). It looks complicated.
